### DISCUSSION TITLE
Relax `CS4012` on ROS with the scoped keyword

### DISCUSSION THREAD

**Discussion by RenderMichael** [dotnet/csharplang#6786] [Language Ideas] [answered]
Currently, the line
``` csharp
async void MyMethod()
{
    ReadOnlySpan<char> sessionId = "myString".AsSpan();
}
```
throws a compiler error `CS4012: Parameters or locals of type 'ReadOnlySpan<char>' cannot be declared in async methods or lambda expressions`

I propose that this be relaxed with the `scoped` keyword
``` csharp
async void MyMethod()
{
    scoped ReadOnlySpan<char> sessionId = "myString".AsSpan();
}
```

**Comment by svick** [comment, marked as answer]
It's not clear to me what is the proposed behavior when it comes to crossing `await`s or what is the reason for the `scoped` keyword, but this looks like a duplicate of https://github.com/dotnet/csharplang/discussions/4687.

**Reply by RenderMichael** [reply]
This is for the happy-path scenario of uncrossed awaits.

Being accessed on the other side of an await would force it onto the heap, using `scoped` here overloads the concept to refer to "not accessed outside of this region". Much like how `scoped` is normally used, it's just to ensure the span is being used properly. But it is strictly unnecessary.

Thanks for pointing me to #4687, except for my proposed solution this is a duplicate
